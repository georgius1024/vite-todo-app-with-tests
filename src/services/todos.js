export default {
  counter: 0,
  list: [],
  load() {
    this.list = ["item 1", "item 2", "item 3"].map((text, id) => ({
      id,
      text,
    }))
  },
  add(item) {
    this.counter++;
    this.list.push({ ...item, id: this.counter });
    return this.counter;
  },
  update(id, updated) {
    this.list = this.list.map((item) => {
      if (item.id === id) {
        return { ...updated, id };
      }
      return item;
    });
  },
  destroy(id) {
    this.list = this.list.filter((item) => item.id !== id);
  },
};
