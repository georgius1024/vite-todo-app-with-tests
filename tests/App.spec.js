import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";
import Todos from "src/services/todos.js";
import component from "src/App.vue";
import todos from "../src/services/todos.js";

describe("App component", () => {
  beforeEach(() => {
    todos.load = vi.fn().mockImplementation(() => {
      todos.list = ["item 1", "item 2", "item 3", "item 4"].map((text, id) => ({
        id,
        text,
      }));
    });
    todos.add = vi.fn();
    todos.destroy = vi.fn();
  });
  afterEach(() => {
    vi.resetAllMocks();
  });
  it("mounts without error", () => {
    const wrapper = mount(component);
    expect(wrapper).toBeDefined();
    expect(todos.load).toHaveBeenCalled();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders items list", () => {
    const wrapper = mount(component);
    const li = wrapper.findAll("li");
    expect(li).toHaveLength(4);
  });

  it("click on remove control reduces number of items", async () => {
    const wrapper = mount(component);
    const removeControl = wrapper.find('[data-testid="item-remove-0"]');
    await removeControl.trigger("click");
    expect(todos.destroy).toHaveBeenCalled();
  });

  it("can add new item to list", async () => {
    const wrapper = mount(component);
    const inputControl = wrapper.find('[data-testid="item-input"]');
    const buttonControl = wrapper.find('[data-testid="item-submit"]');
    await inputControl.setValue("new item");
    await buttonControl.trigger("click");
    expect(todos.add).toHaveBeenCalled();
  });
});
