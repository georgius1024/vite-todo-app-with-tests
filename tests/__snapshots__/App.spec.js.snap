// Vitest Snapshot v1

exports[`App component > mounts without error 1`] = `
"<div data-v-7a7a37b1=\\"\\" class=\\"app\\">
  <h1 data-v-7a7a37b1=\\"\\">TodoApp</h1>
  <ul data-v-4afa86a5=\\"\\" data-v-7a7a37b1=\\"\\" class=\\"todo\\" data-testid=\\"item-list\\">
    <li data-v-4afa86a5=\\"\\" data-testid=\\"item-0\\"><span data-v-4afa86a5=\\"\\" class=\\"remove\\" data-testid=\\"item-remove-0\\">×</span> item 1</li>
    <li data-v-4afa86a5=\\"\\" data-testid=\\"item-1\\"><span data-v-4afa86a5=\\"\\" class=\\"remove\\" data-testid=\\"item-remove-1\\">×</span> item 2</li>
    <li data-v-4afa86a5=\\"\\" data-testid=\\"item-2\\"><span data-v-4afa86a5=\\"\\" class=\\"remove\\" data-testid=\\"item-remove-2\\">×</span> item 3</li>
    <li data-v-4afa86a5=\\"\\" data-testid=\\"item-3\\"><span data-v-4afa86a5=\\"\\" class=\\"remove\\" data-testid=\\"item-remove-3\\">×</span> item 4</li>
  </ul>
  <hr data-v-7a7a37b1=\\"\\">
  <div data-v-2bcb7aa7=\\"\\" data-v-7a7a37b1=\\"\\" class=\\"form\\"><input data-v-2bcb7aa7=\\"\\" placeholder=\\"enter todo\\" data-testid=\\"item-input\\"><button data-v-2bcb7aa7=\\"\\" data-testid=\\"item-submit\\">Add</button></div>
</div>"
`;
